Readme.md

# stripe-cohort

  Create [Stripe](https://stripe.com) customer cohorts for a useful business overview.

### Overview

If you're interested in getting a cohort based overview of your Stripe customers, then this library might be of help.

It currently supports the following stats (by cohort):

- Total number of customers
- Monthly recurring revenue 
- Number of subscriptions
- Subscriptions by plan

## Installation

    $ npm install stripe-cohort

## Example

Create a new Stripe cohort (by the customer's `created` date):

```js
cohort(new Date('1/1/2014'), new Date('2/1/2014'), function (err, customers) {
  console.log(customers.count() + ' created in January!');
});
```

And the returned `customers` object lets you dive deeper into the cohort.

#### Number of Customers

You can query the total amount of customers returned:

```js
customers.count();
```

or filter further inside the cohort by the customers' `created` date:

```js
customers.count(new Date('1/15/2014'), new Date('1/24/2014'));
```

#### Customer List

```js
customers.list()
```
```js
[
  {
    id: 'cus_2983jd92d2d',
    name: 'Patrick Collison',
    ..
  },
]
```
or filter further by the customers' `created` date:

```js
customers.list(new Date('1/15/2014'), new Date('1/24/2014'));
```

or get all the `delinquent` customers:

```js
customers.delinquent().count();
```

#### Subscriptions

You can learn about your active subscriptions too:

```js
customers.subscriptions().count();
```

Or the ones created between the date provided:

```js
customers.subscriptions(new Date('1/15/2014'), new Date('1/24/2014')).count();
```

Or just get the list of Stripe subscription objects:

```js
var objects = customers.subscriptions().list();
```

#### Monthly Recurring Revenue

You can get the monthly recurring revenue from the active subscriptions on the customers:

```js
customers.subscriptions().active().mrr();
```

And for the trialing accounts:

```js
customers.subscriptions().trialing().mrr();
```

And for any status really:

```js
customers.subscriptions().status('unpaid').mrr();
```

And you can query the monthly recurring revenue  by subscription `start` within a cohort: 

```js
customers.subscriptions().active().mrr(new Date('1/15/2014'), new Date('1/16/2014'));
```

Remember that the montly recurring revenue does not equal charges. For example, if a customer upgrades from a $29 plan to a $79 plan today, they will pro-rated for the rest of their billing period. That means you did not make the $79 yet, but you'll make the difference next month. For hard cash, use [stripe-charges](https://github.com/segmentio/stripe-charges).

#### Plans

It's also interesting to know what plans the subscriptions are being set at. You can select the subscriptions that fall under that plan:

```js
var mrr = customers.subscriptions().active().plan('startup').mrr();
console.log('We made $' + mrr + ' off the startup plan!');
```

## License

MIT

History.md
0.0.3 - February 26, 2014
-------------------------
Normalizing subscriptions for MRR

0.0.1 - February 26, 2014
-------------------------
:sparkles:


test/index.js

var assert = require('assert');
var cohort = require('..');
var util = require('util');

describe('stripe-cohort', function () {

  // TODO: enter your stripe key
  var key = 'stripe-key';

  describe('#cohort', function () {
    this.timeout(30000); // querying lots of customers can take a while

    it('should return a cohort of customers', function (done) {
      var self = this;
      var start = new Date('2/1/2014');
      var end = new Date('3/1/2014');
      cohort(key)(start, end, function (err, customers) {
        if (err) return done(err);
        assert(customers);
        self.customers = customers;
        done();
      });
    });
  });

  describe('#customers', function () {

    describe('#count', function () {
      it('should count the customers', function () {
        var count = this.customers.count();
        assert('number' === typeof count);
      });
    });

    describe('#list', function () {
      it('should get a list of customers', function () {
        var list = this.customers.list();
        assert(util.isArray(list));
      });

      it('should get a list of customers by created date', function () {
        var start = new Date('2/15/2014');
        var end = new Date('2/16/2014');
        var list = this.customers.list(start, end);
        assert(util.isArray(list));
      });
    });
  });


  describe('#subscriptions', function () {

    describe('#count', function () {
      it('should count the customers', function () {
        var count = this.customers.subscriptions().count();
        assert('number' === typeof count);
      });
    });

    describe('#list', function () {
      it('should get a list of subscriptions', function () {
        var list = this.customers.subscriptions().list();
        assert(util.isArray(list));
      });

      it('should get a list of subscriptions by started date', function () {
        var start = new Date('2/15/2014');
        var end = new Date('2/16/2014');
        var list = this.customers.subscriptions(start, end).list();
        assert(util.isArray(list));
      });
    });

    describe('#mrr', function () {
      it('should calculate the monthly recurring revenue of all subscription', function () {
        var mrr = this.customers.subscriptions().mrr();
        assert('number' === typeof mrr);
      });
    });

    describe('#active', function () {
      it('should calculate the monthly recurring revenue of all active subscription', function () {
        var mrr = this.customers.subscriptions().active().mrr();
        assert('number' === typeof mrr);
      });

      it('should calculate the monthly recurring revenue of all active subscription by started date', function () {
        var start = new Date('2/15/2014');
        var end = new Date('2/16/2014');
        var mrr = this.customers.subscriptions().active().mrr(start, end);
        assert('number' === typeof mrr);
      });
    });

    describe('#trialing', function () {
      it('should calculate the monthly recurring revenue of all trialing subscription', function () {
        var mrr = this.customers.subscriptions().trialing().mrr();
        assert('number' === typeof mrr);
      });
    });
  });
});

lib/index.js

var Batch = require('batch');
var Customers = require('./customers');
var debug = require('debug')('stripe-cohort');
var defaults = require('defaults');
var range = require('range-component');
var Stripe = require('stripe');
var unixTime = require('unix-time');
var util = require('util');

/**
 * Expose `Cohorter`.
 */

module.exports = Cohorter;

/**
 * Initialize a new `Cohorter` the Stripe `key`.
 *
 * @param {String} key
 * @param {Object} options
 *   @param {Number} count
 */

function Cohorter (key, options) {
  if (!(this instanceof Cohorter)) return new Cohorter(key, options);
  if (!key) throw new Error('Stripe cohort requires a Stripe key.');
  this.stripe = Stripe(key);
  this.options = defaults(options, { count: 100, concurrency: 1 });
  var self = this;
  return function () { self.cohort.apply(self, arguments); };
}

/**
 * Create a cohort.
 *
 * @param {Date} start
 * @param {Date} end
 * @param {Function} callback
 */

Cohorter.prototype.cohort = function (start, end, callback) {
  if (!util.isDate(start)) throw new Error('Start must be a date.');
  if (!util.isDate(end)) throw new Error('End must be a date.');
  var self = this;
  debug('creating cohort [%s - %s] ..', start, end);
  this.customers(start, end, function (err, customers) {
    if (err) return callback(err);
    debug('created cohort');
    callback(null, new Customers(customers));
  });
};

/**
 * Load customers between a `start` and `end` date.
 *
 * @param {Date} start
 * @param {Date} end
 * @param {Function} callback
 */

Cohorter.prototype.customers = function (start, end, callback) {
  debug('loading customers with created [%s - %s] ..', start, end);
  var self = this;
  var customers = [];
  var page = this.options.count;
  // run the first query to get the total unpaginated count
  this.query(start, end, 0, function (err, res) {
    if (err) return callback(err);
    var count = res.count;
    customers.push.apply(customers, res.data);
    var got = res.data.length;
    var left = res.count - got;
    // check if we grabbed everything in the first query
    if (0 === left) return callback(null, customers);
    // there's more, we have to paginate query
    var pages = Math.ceil(left / page);
    debug('loaded %d customers, %d left in %d pages of %d', got, left, pages, page);
    var batch = new Batch();
    batch.concurrency(self.options.concurrency);
    range(0, pages).forEach(function (i) {
      batch.push(function (done) { self.query(start, end, got + (i * page), done); });
    });
    batch.end(function (err, results) {
      if (err) return callback(err);
      results.forEach(function (res) {
        customers.push.apply(customers, res.data);
      });
      debug('finished loading all customers in cohort');
      callback(null, customers);
    });
  });
};

/**
 * List customers between a `start` and `end` date, with an `offset`.
 *
 * @param {Date} start
 * @param {Date} end
 * @param {Number} offset
 * @param {Function} callback
 */

Cohorter.prototype.query = function (start, end, offset, callback) {
  debug('loading customers with created [%s - %s] with offset %d ..', start, end, offset);
  var options = {
    created: { gte: unixTime(start), lte: unixTime(end) },
    count: this.options.count,
    offset: offset
  };
  this.stripe.customers.list(options, function (err, res) {
    if (err) return callback(err);
    var customers = res.data;
    debug('loaded %d customers with offset %d', customers.length, offset);
    callback(null, res);
  });
};

lib/customers.js

var Subscriptions = require('./subscriptions');
var util = require('util');

/**
 * Expose `Customers`.
 */

module.exports = Customers;

/**
 * A cohort of Stripe `customers`.
 *
 * @param {Array|Customer} customers
 *
 */

function Customers (customers) {
  this.customers = customers;
}

/**
 * Return a new `Customers` thats filtered by `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Customers}
 */

Customers.prototype.created = function (start, end) {
  if (!util.isDate(start)) return this;
  if (!util.isDate(end)) end = new Date('1/1/99999');
  var s = start.getTime(), e = end.getTime();
  return new Customers(this.customers.filter(function (customer) {
    var created = customer.created * 1000;
    return created >= s && created <= e;
  }));
};

/**
 * Return a new `Customers` thats filtered by delinquent customers.
 *
 * @param {Boolean} delinquent
 * @return {Customers}
 */

Customers.prototype.delinquent = function (delinquent) {
  if (typeof delinquent !== 'boolean') delinquent = true;
  return new Customers(this.customers.filter(function (customer) {
    return customer.delinquent === delinquent;
  }));
};

/**
 * Filter customers by `fn`.
 *
 * @param {Function} fn
 * @return {Customers}
 */

Customers.prototype.filter = function (fn) {
  return new Customers(this.customers.filter(fn));
};

/**
 * Count the customers between `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Number}
 */

Customers.prototype.list = function (start, end) {
  return this.created(start, end).customers;
};

/**
 * Count the customers between `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Number}
 */

Customers.prototype.count = function (start, end) {
  return this.created(start, end).customers.length;
};

/**
 * Get a list of active subscriptions between `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Subscriptions}
 */

Customers.prototype.subscriptions = function (start, end) {
  var all = [];
  this.customers.forEach(function (customer) {
    var subscriptions = customer.subscriptions.data;
    subscriptions.forEach(function (subscription) {
      // create a backwards reference so that we can check discount
      subscription.customer = customer;
    });
    all.push.apply(all, subscriptions);
  });
  return new Subscriptions(all).started(start, end);
};

lib/subscriptions.js
var util = require('util');

/**
 * Expose `Subscriptions`.
 */

module.exports = Subscriptions;

/**
 * A list of Stripe `subscriptions`.
 *
 * @param {Array|Customer} subscriptions
 *
 */

function Subscriptions (subscriptions) {
  this.subscriptions = subscriptions;
}

/**
 * Return a new active `Subscriptions` list thats filtered by `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Subscriptions}
 */

Subscriptions.prototype.active = function (start, end) {
  return this.started(start, end).status('active');
};

/**
 * Return a new active `Subscriptions` list thats filtered by `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Subscriptions}
 */

Subscriptions.prototype.trialing = function (start, end) {
  return this.started(start, end).status('trialing');
};

/**
 * Return a new `Subscriptions` list thats filtered by `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Subscriptions}
 */

Subscriptions.prototype.started = function (start, end) {
  if (!util.isDate(start)) return this;
  if (!util.isDate(end)) end = new Date('1/1/99999');
  var s = start.getTime(), e = end.getTime();
  return new Subscriptions(this.subscriptions.filter(function (subscription) {
    if (subscription.status !== 'active') return false;
    var started = subscription.start * 1000;
    return started >= s && started <= e;
  }));
};

/**
 * Return a new `Subscriptions` list thats filtered by `status`.
 *
 * @param {String} status
 * @return {Subscriptions}
 */

Subscriptions.prototype.status = function (status) {
  return new Subscriptions(this.subscriptions.filter(function (subscription) {
    return subscription.status === status;
  }));
};

/**
 * Return a new `Subscriptions` list thats filtered by `planId`.
 *
 * @param {String} planId
 * @return {Subscriptions}
 */

Subscriptions.prototype.plan = function (planId) {
  return new Subscriptions(this.subscriptions.filter(function (subscription) {
    return subscription.plan.id === planId;
  }));
};

/**
 * Return a new `Subscriptions` list thats filtered without
 * plans with `planId`.
 *
 * @param {String} planId
 * @return {Subscriptions}
 */

Subscriptions.prototype.withoutPlan = function (planId) {
  return new Subscriptions(this.subscriptions.filter(function (subscription) {
    return subscription.plan.id !== planId;
  }));
};

/**
 * Return a new `Subscriptions` list thats has plans that cost more
 * or equal to `amount`.
 *
 * @param {Number} amount
 * @return {Subscriptions}
 */

Subscriptions.prototype.paid = function (amount) {
  if (!amount) amount = 1;
  return new Subscriptions(this.subscriptions.filter(function (subscription) {
    return subscription.plan.amount >= amount;
  }));
};

/**
 * Count the subscriptions between `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Number}
 */

Subscriptions.prototype.list = function (start, end) {
  return this.started(start, end).subscriptions;
};

/**
 * Count the subscriptions between `start` and `end`.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Number}
 */

Subscriptions.prototype.count = function (start, end) {
  return this.started(start, end).subscriptions.length;
};

/**
 * Print an audit of the subscriptions.
 *
 * @return {Number}
 */

Subscriptions.prototype.print = function () {
  var list = this.list();
  var total = 0;
  for (var i = 0; i < list.length; i += 1) {
    var s = list[i];
    var c = s.customer;
    var a = Math.round(amount(s) * 100) / 100.0;
    console.log([c.email, s.status, s.plan.name, '$' + a].join(' - '));
    total += a;
  }
  console.log('Total MRR: $' + total);
};

/**
 * Count the monthly recurring revenue of subscriptions in that period.
 *
 * @param {Date} start
 * @param {Date} end
 * @return {Number}
 */

Subscriptions.prototype.mrr = function (start, end) {
  var mrr = this.started(start, end).subscriptions.reduce(function (memo, subscription) {
    return memo + amountMrr(subscription);
  }, 0.00);

  return Math.round(mrr * 100) / 100.0; // to two decimal points
};

/**
 * Calculate the amount in dollars of a subscription after fees
 * and discounts.
 *
 * @param {Subscription} subscription
 * @return {Number}
 */

function amountMrr (subscription) {
  var res = (subscription.plan.amount / 100.0);
  res = normalizeMonths(subscription, res);
  if (res > 0.0) {
    var discount = subscription.customer.discount;
    if (discount && discount.coupon) {
      var coupon = discount.coupon;
      if (coupon.valid) {
        if (coupon.amount_off) res -= (coupon.amount_off / 100.0);
        else if (coupon.percent_off) res *= (1.00 - (coupon.percent_off / 100.0));
      }
    }
    res *= (1.00 - 0.029); // Stripe fees
  }
  return res;
}

//
function normalizeMonths (subscription, res) {
  var interval = subscription.plan.interval;
  var count = subscription.plan.interval_count;
  if (interval === 'day')
    res *= 30 / count;
  else if (interval === 'week')
    res *= 4 / count;
  else if (interval === 'month')
    res *= 1 / count;
  else if (interval === 'year'){
    res = res / 12 / count;
  }
  else
    throw new Error('Unexpected interval ' + interval);
  return res;
}
